ecosystem/indexer-grpc/indexer-transaction-generator/src/transaction_generator.rs

```diff
@@ -76,12 +97,185 @@ impl TransactionGenerator {
         self.managed_node.stop().await
     }
 
+    /// Go through the test case steps and generate the transactions.
     async fn generate(&self, test_case: TestCase) -> anyhow::Result<()> {
-        // TODO: finish the test generation logic.
         println!(
-            "Generate transaction for test case {} at {:?} ",
+            "\n Generate transaction for test case {} at {:?} \n",
             test_case.name, self.output_folder,
         );
+        // Test case output folder.
+        let test_case_output_folder = self.output_folder.join(&test_case.name);
+        // If the output folder doesn't exist, create it.
+        if !test_case_output_folder.exists() {
+            std::fs::create_dir_all(&test_case_output_folder)?;
+        }
+        let mut transactions_version_to_capture = vec![];
+        // Execute the transactions.
+        for step in test_case.steps {
+            match step {
+                Step::Setup((path, _, config)) => {
+                    // Change current directory to the setup script folder.
+                    std::env::set_current_dir(&path)
+                        .context(format!("Failed to change directory to: {:?}", path))?;
+                    if let Some(test_config) = config {
+                        if let Some(amount) = test_config.fund_amount {
+                            println!("Funding account with amount: {}", amount);
+                            let fund = FundWithFaucet::new_for_indexer_testing(amount);
+                            fund.execute()
+                                .await
+                                .context("Failed to fund the account.")?;
+                            println!("Funded account with amount: {}", amount);
+                        }
+                    }
+                    println!("Compiling and running the setup script: {:?}", path);
+                    // Compile the setup script.
+                    let cmd =
+                        aptos::move_tool::CompilePackage::new_for_indexer_testing(path.clone());
+                    let _ = cmd
+                        .execute()
+                        .await
+                        .context(format!("Failed to compile package: {:?}", path))?;
+                    println!("Compiled the setup script: {:?}", path);
+                    // Publish the setup script.
+                    let cmd =
+                        aptos::move_tool::PublishPackage::new_for_indexer_testing(path.clone());
+                    let _ = cmd
+                        .execute()
+                        .await
+                        .context(format!("Failed to publish module: {:?}", path))?;
+                },
+                Step::Action((path, _)) => {
```

Comment:
Just an idea, theoretically all of this stuff you do here could instead be a method on `Step`. That way you don't have to make as much of Step public, you'd just do like `std::env::set_current_dir(self.path)`. It'd also help make helpers for common things. e.g. it seems like it'd break everything if you forgot to call `std::env::set_current_dir` for every step variant.